```
--- FAIL: TestCheckGCSWeakBuilds (0.61s)
	e2e_test.go:464: expected: false, saw: true
	e2e_test.go:467: expected: map[foo:{Not Stable 42} bar:{Stable 44}], saw: map[foo:{Stable 42} bar:{Stable 44}]
```

--- a/mungegithub/mungers/e2e/e2e.go
+++ b/mungegithub/mungers/e2e/e2e.go
@@ -219,19 +219,19 @@ func (e *RealE2ETester) GCSWeakStable() bool {
 
 		// If we're here it means that we weren't able to find a test that failed, which means that the reason of build failure is comming from the infrastructure
 		// Check results of previous two builds.
+		unstable := make([]int, 0)
 		if stable, err := e.GoogleGCSBucketUtils.CheckFinishedStatus(job, lastBuildNumber-1); !stable || err != nil {
-			e.setBuildStatus(job, "Not Stable", strconv.Itoa(lastBuildNumber))
-			allStable = false
-			glog.Infof("WeakStable failed because found a weak failure in build %v and build %v failed.", lastBuildNumber, lastBuildNumber-1)
-			continue
+			unstable = append(unstable, lastBuildNumber-1)
 		}
 		if stable, err := e.GoogleGCSBucketUtils.CheckFinishedStatus(job, lastBuildNumber-2); !stable || err != nil {
+			unstable = append(unstable, lastBuildNumber-2)
+		}
+		if len(unstable) > 1 {
 			e.setBuildStatus(job, "Not Stable", strconv.Itoa(lastBuildNumber))
 			allStable = false
-			glog.Infof("WeakStable failed because found a weak failure in build %v and build %v failed.", lastBuildNumber, lastBuildNumber-2)
+			glog.Infof("WeakStable failed because found a weak failure in build %v and builds %v failed.", lastBuildNumber, unstable)
 			continue
 		}
-
 		e.setBuildStatus(job, "Stable", strconv.Itoa(lastBuildNumber))
 	}
 	return allStable

--- a/mungegithub/mungers/e2e/e2e_test.go
+++ b/mungegithub/mungers/e2e/e2e_test.go
@@ -331,7 +331,7 @@ func TestCheckGCSWeakBuilds(t *testing.T) {
 			},
 		},
 		// If the last build was unsuccessful but there's no failures in JUnit file we assume that it was
-		// an infrastructure failure. Build should succeed if previous two builds were fully successful.
+		// an infrastructure failure. Build should succeed if at least one of two builds were fully successful.
 		{
 			paths: map[string][]byte{
 				"/foo/latest-build.txt": []byte(strconv.Itoa(latestBuildNumberFoo)),
@@ -341,7 +341,7 @@ func TestCheckGCSWeakBuilds(t *testing.T) {
 				}, t),
 				fmt.Sprintf("/foo/%v/artifacts/junit_01.xml", latestBuildNumberFoo): getJUnit(5, 0),
 				fmt.Sprintf("/foo/%v/finished.json", latestBuildNumberFoo-1): marshalOrDie(utils.FinishedFile{
-					Result:    "SUCCESS",
+					Result:    "UNSTABLE",
 					Timestamp: 1233,
 				}, t),
 				fmt.Sprintf("/foo/%v/finished.json", latestBuildNumberFoo-2): marshalOrDie(utils.FinishedFile{
@@ -361,7 +361,7 @@ func TestCheckGCSWeakBuilds(t *testing.T) {
 			},
 		},
 		// If the last build was unsuccessful but there's no failures in JUnit file we assume that it was
-		// an infrastructure failure. Build should fail if of of two builds failed.
+		// an infrastructure failure. Build should fail more than both recent builds failed.
 		{
 			paths: map[string][]byte{
 				"/foo/latest-build.txt": []byte(strconv.Itoa(latestBuildNumberFoo)),
@@ -371,7 +371,7 @@ func TestCheckGCSWeakBuilds(t *testing.T) {
 				}, t),
 				fmt.Sprintf("/foo/%v/artifacts/junit_01.xml", latestBuildNumberFoo): getJUnit(5, 0),
 				fmt.Sprintf("/foo/%v/finished.json", latestBuildNumberFoo-1): marshalOrDie(utils.FinishedFile{
-					Result:    "SUCCESS",
+					Result:    "UNSTABLE",
 					Timestamp: 1233,
 				}, t),
 				fmt.Sprintf("/foo/%v/finished.json", latestBuildNumberFoo-2): marshalOrDie(utils.FinishedFile{
@@ -390,7 +390,7 @@ func TestCheckGCSWeakBuilds(t *testing.T) {
 				"bar": {Status: "Stable", ID: "44"},
 			},
 		},
-		// If the last build was unsuccessful and theres a failed test in a JUnit file we should fail.
+		// If the last build was unsuccessful and there's a failed test in a JUnit file we should fail.
 		{
 			paths: map[string][]byte{
 				"/foo/latest-build.txt": []byte(strconv.Itoa(latestBuildNumberFoo)),
@@ -428,7 +428,7 @@ func TestCheckGCSWeakBuilds(t *testing.T) {
 				}, t),
 				fmt.Sprintf("/bar/%v/artifacts/junit_01.xml", latestBuildNumberBar): getJUnit(5, 0),
 				fmt.Sprintf("/bar/%v/artifacts/junit_02.xml", latestBuildNumberBar): getJUnit(5, 1),
-				fmt.Sprintf("/bar/%v/artifacts/junit_03.xml", latestBuildNumberBar): getJUnit(5, 0),
+				fmt.Sprintf("/bar/%v/artifacts/junit_03.xml", latestBuildNumberBar): getJUnit(5, 1),
 			},
 			expectStable: false,
 			expectedStatus: map[string]BuildInfo{

